kernel/thread/thread.c

```diff
@@ -565,6 +565,11 @@ int thd_destroy(kthread_t *thd) {
        and unblock them. */
     genwait_wake_all(thd);
 
+    /* If this thread was waiting on anything, remove it from
+       the genwait queue. */
+    if(thd->wait_obj)
+        genwait_wake_thd(thd->wait_obj, thd, 0);
```

Comment:
If a user were to call `thd_destroy` on a thread that is waiting, it would otherwise not have been removed from the waiting queue and be returned as freed memory onto the runnable list. This also ensures that the genwait list is left clean after `thd_shutdown`